packages/conventional-changelog/index.js

```diff
@@ -8,7 +8,23 @@ function conventionalChangelog (options, context, gitRawCommitsOpts, parserOpts,
 
   if (options.preset) {
     try {
-      options.config = conventionalChangelogPresetLoader(options.preset.toLowerCase())
+      let presetConfig = null
+      if (typeof options.preset === 'object' && options.preset.name) {
+        // Rather than a string preset name, options.preset can be an object
+        // with a "name" key indicating the preset to load; additinoal key/value
+        // pairs are assumed to be configuration for the preset. See the documentation
+        // for a given preset for configuration available.
+        presetConfig = options.preset
+        options.config = conventionalChangelogPresetLoader(options.preset.name.toLowerCase())
+      } else {
+        options.config = conventionalChangelogPresetLoader(options.preset.toLowerCase())
+      }
+      // rather than returning a promise, presets can return a builder function
+      // which accepts a config object (allowing for customization) and returns
+      // a promise.
+      if (!options.config.then && presetConfig) {
+        options.config = options.config(presetConfig)
```

Comment:
I _believe_ this is redundant. `conventional-changelog-core` currently merges the options from, say, `parserOpts`, over the `options.config.parserOpts` object. Same with `gitRawCommitOpts` and `writerOpts`.

If you don't like the preset options, you can override them by settings the appropriate options in `parserOpts`, etc.

Reply:
I honestly find this approach to configuring things very convoluted, it was @stevemao's recommendation that we start passing configuration directly into presets, and I think this is much less confusing to an implementor.

Reply:
also, I think the configuration is at the wrong level of granularity; as evidenced by the variety of open issues on `standard-version` (which is the project I work with most directly) users want to be able to:

* have more control over the URL format output in CHANGELOGS.
* have more control over which specific types of commits show up in the CHANGELOG (`feat`, vs, `chore`, etc).

The existing configuration exposes a ton of nuts and bolts about what's happening in conventional-commits several layers of abstraction away from the presets; as an example, even though you could potentially overwrite the entire `writeOpts` config, there's no easy way prior to this change to pass configuration to [this method](https://github.com/conventional-changelog/conventional-changelog/blob/master/packages/conventional-changelog-angular/writer-opts.js#L36) in the angular preset, which actually configures which items are toggled oin and off in the CHANGELOG.

Reply:
@stevemao is there a better way to pass configuration all the way through to the preset? without introducing this refactor -- I want the preset to have access to configuration like you suggested, but I think it should be the responsibility of upstream libraries like `standard-version` to accept this configuration (either from command line arguments or loading it from disk).

Long story short, the least breaking way I could think of to chain configuration all the way to the preset was to override the behavior such that `presets` can optionally return a factory that accepts configuration, rather than a promise.

Reply:
> The existing configuration exposes a ton of nuts and bolts about what's happening in conventional-commits several layers of abstraction away from the presets.

Thank you for the explanation @bcoe. :+1:  It's a sentiment I share as well. :sweat_smile:

Reply:
Is there an example on how to make angular preset print each commit type in the change log? Or do I still have to implement a workaround [like this](https://github.com/saltstack-formulas/template-formula/blob/c71915ed1e38d72c38360d409ddcffd01a61f557/release.config.js#L23-L104)? My apologies, but I am somewhat confused. This PR is merged, but the `master` still prints only 3 types be default - `feat`, `fix` and `perf`.